src/main/java/org/zalando/nakadi/service/job/TimelineCleanupJob.java

```diff
@@ -0,0 +1,81 @@
+package org.zalando.nakadi.service.job;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.scheduling.annotation.Scheduled;
+import org.springframework.stereotype.Service;
+import org.zalando.nakadi.domain.Timeline;
+import org.zalando.nakadi.exceptions.TopicDeletionException;
+import org.zalando.nakadi.exceptions.runtime.InconsistentStateException;
+import org.zalando.nakadi.exceptions.runtime.RepositoryProblemException;
+import org.zalando.nakadi.repository.TopicRepository;
+import org.zalando.nakadi.repository.db.EventTypeCache;
+import org.zalando.nakadi.repository.db.TimelineDbRepository;
+import org.zalando.nakadi.service.timeline.TimelineService;
+
+@Service
+public class TimelineCleanupJob {
+
+    private static final String JOB_NAME = "timelines-cleanup";
+
+    private static final Logger LOG = LoggerFactory.getLogger(TimelineCleanupJob.class);
+
+    private final EventTypeCache eventTypeCache;
+    private final TimelineDbRepository timelineDbRepository;
+    private final TimelineService timelineService;
+    private final ExclusiveJobWrapper jobWrapper;
+
+    @Autowired
+    public TimelineCleanupJob(final EventTypeCache eventTypeCache,
+                              final TimelineDbRepository timelineDbRepository,
+                              final TimelineService timelineService,
+                              final JobWrapperFactory jobWrapperFactory,
+                              @Value("${nakadi.jobs.timelineCleanup.runPeriodMs}") final int periodMs) {
+        this.eventTypeCache = eventTypeCache;
+        this.timelineDbRepository = timelineDbRepository;
+        this.timelineService = timelineService;
+        this.jobWrapper = jobWrapperFactory.createExclusiveJobWrapper(JOB_NAME, periodMs);
+    }
+
+    @Scheduled(
+            fixedDelayString = "${nakadi.jobs.checkRunMs}",
+            initialDelayString = "${random.int(${nakadi.jobs.checkRunMs})}")
+    public void cleanupTimelines() {
+        try {
+            jobWrapper.runJobLocked(() ->
+                    timelineDbRepository.getExpiredTimelines().stream()
+                            .forEach(timeline -> {
+                                deleteTimelineTopic(timeline);
```

Comment:
1) If the topic deletion failed and timeline is marked as deleted - the topic will "hang in air"
2) If the topic is deleted and the timeline is failed to be marked as deleted - then it will try again the next time Job runs.

It seems that #1 can be improved but I don't really see how. In general we can change it so that if topic failed to be removed - we don't mark timeline as deleted - so that it will try the next time. But the problem can be that the topic failed to be removed because it doesn't exist (e.g. was removed by hand or whatever), then in this case the timeline will try to be cleaned each Job run (so it will get into this infinite loop).

Do you see any ways to improve that?